File: experiments/014-LiveTypingFiveRuns/cells/6-LiveTypingRefactoring_2-EvaluateAsLastResource+5-LiveTyping+6-LiveTypingRefactoring+8-DesignHeuristicsInline_free/2024-1c-parcial1/20260917-024928-14557/exercise/java/src/main/java/simshotels/Hotel.java
package simshotels;

import java.util.ArrayList;

public class Hotel {

    private ArrayList<Floor> floors;
    private int availableRoomsCount;

    // initialization

    public Hotel() {
        availableRoomsCount = 0;
    }

    // accounting

    public int totalLosses() {
        int acc;

        acc = 0;
        for (Floor floor : floors) {
            acc = acc + floor.totalLosses();
        }

        return acc;
    }

    public int totalProfits() {
        int acc;

        acc = 0;
        for (Floor floor : floors) {
            acc = acc + floor.totalProfits();
        }

        return acc;
    }

    // setters

    public Hotel setFloors(ArrayList<Floor> aFloorsCollection) {
        Hotel.assertHaveFloors(aFloorsCollection);

        floors = aFloorsCollection;

        availableRoomsCount = totalRooms();

        return this;
    }

    // totals

    public int totalRooms() {
        int ix;
        int acc;

        acc = 0;

        ix = 1;
        while (ix <= floors.size()) {
            Floor floor;
            int jx;
            floor = floors.get(ix - 1);

            jx = 1;
            while (jx <= floor.rooms().size()) {
                acc = acc + 1;

                jx = jx + 1;
            }

            ix = ix + 1;
        }

        return acc;
    }

    public int totalRoomsAvailable() {
        return availableRoomsCount;
    }

    public int totalRoomsOccupied() {
        int ix;
        int acc;

        acc = 0;

        ix = 1;
        while (ix <= floors.size()) {
            Floor floor;
            int jx;
            floor = floors.get(ix - 1);

            jx = 1;
            while (jx <= floor.rooms().size()) {
                Room room;
                room = floor.rooms().get(jx - 1);
                if (room.isOccupied()) {
                    acc = acc + 1;
                }

                jx = jx + 1;
            }

            ix = ix + 1;
        }

        return acc;
    }

    public int totalRoomsReserved() {
        int ix;
        int acc;

        acc = 0;

        ix = 1;
        while (ix <= floors.size()) {
            Floor floor;
            int jx;
            floor = floors.get(ix - 1);

            jx = 1;
            while (jx <= floor.rooms().size()) {
                Room room;
                room = floor.rooms().get(jx - 1);
                if (room.isReserved()) {
                    acc = acc + 1;
                }

                jx = jx + 1;
            }

            ix = ix + 1;
        }

        return acc;
    }

    // testing

    public boolean isEmpty() {
        int ix;
        boolean acc;

        acc = true;

        ix = 1;
        while (ix <= floors.size()) {
            Floor floor;
            int jx;
            floor = floors.get(ix - 1);

            jx = 1;
            while (jx <= floor.rooms().size()) {
                Room room;
                room = floor.rooms().get(jx - 1);
                acc = acc && room.isAvailable();

                jx = jx + 1;
            }

            ix = ix + 1;
        }

        return acc;
    }

    // guests

    public void receiveAtFloorAtRoom(String aGuestType, int aFloorNumber, int aRoomNumber) {
        Floor floor;
        Room room;

        if (aFloorNumber < 1 || aFloorNumber > floors.size()) {
            Hotel.signalFloorNumberDoesNotExist();
        }
        floor = floors.get(aFloorNumber - 1);
        if (aRoomNumber < 1 || aRoomNumber > floor.rooms().size()) {
            Hotel.signalRoomNumberDoesNotExist();
        }
        room = floor.rooms().get(aRoomNumber - 1);

        room.receive(aGuestType);

        availableRoomsCount = availableRoomsCount - 1;
    }

    public void receiveWithReservationAtFloorAtRoom(String aGuestType, int aFloorNumber, int aRoomNumber) {
        Floor floor;
        Room room;

        if (aFloorNumber < 1 || aFloorNumber > floors.size()) {
            Hotel.signalFloorNumberDoesNotExist();
        }
        floor = floors.get(aFloorNumber - 1);
        if (aRoomNumber < 1 || aRoomNumber > floor.rooms().size()) {
            Hotel.signalRoomNumberDoesNotExist();
        }
        room = floor.rooms().get(aRoomNumber - 1);

        room.receiveWithReservation(aGuestType);
    }

    public void reserveRoomAtFloor(int aRoomNumber, int aFloorNumber) {
        Floor floor;
        Room room;

        if (aFloorNumber < 1 || aFloorNumber > floors.size()) {
            Hotel.signalFloorNumberDoesNotExist();
        }
        floor = floors.get(aFloorNumber - 1);
        if (aRoomNumber < 1 || aRoomNumber > floor.rooms().size()) {
            Hotel.signalRoomNumberDoesNotExist();
        }
        room = floor.rooms().get(aRoomNumber - 1);

        room.reserve();

        availableRoomsCount = availableRoomsCount - 1;
    }

    // assertions (class side)

    public static void assertHaveFloors(ArrayList<Floor> aFloorsCollection) {
        if (aFloorsCollection.isEmpty()) {
            Hotel.signalNoFloors();
        }
    }

    // error description (class side)

    public static String floorNumberDoesNotExistErrorDescription() {
        return "Floor number does not exist";
    }

    public static String noFloorsErrorDescription() {
        return "Cannot have a Hotel without floors";
    }

    public static String roomNumberDoesNotExistErrorDescription() {
        return "Room number does not exist";
    }

    // exceptions (class side)

    public static void signalFloorNumberDoesNotExist() {
        throw new RuntimeException(Hotel.floorNumberDoesNotExistErrorDescription());
    }

    public static void signalNoFloors() {
        throw new RuntimeException(Hotel.noFloorsErrorDescription());
    }

    public static void signalRoomNumberDoesNotExist() {
        throw new RuntimeException(Hotel.roomNumberDoesNotExistErrorDescription());
    }
}

File: exercises/2024-1c-parcial1/python/src/simshotels/hotel.py
class Hotel:

    # initialization

    def __init__(self):
        self._floors = None
        self._available_rooms_count = 0

    # accounting

    def total_losses(self):
        acc = 0
        for floor in self._floors:
            acc = acc + floor.total_losses()

        return acc

    def total_profits(self):
        acc = 0
        for floor in self._floors:
            acc = acc + floor.total_profits()

        return acc

    # setters

    def set_floors(self, a_floors_collection):
        Hotel.assert_have_floors(a_floors_collection)

        self._floors = a_floors_collection

        self._available_rooms_count = self.total_rooms()

        return self

    # totals

    def total_rooms(self):
        acc = 0

        ix = 1
        while ix <= len(self._floors):
            floor = self._floors[ix - 1]

            jx = 1
            while jx <= len(floor.rooms()):
                acc = acc + 1

                jx = jx + 1

            ix = ix + 1

        return acc

    def total_rooms_available(self):
        return self._available_rooms_count

    def total_rooms_occupied(self):
        acc = 0

        ix = 1
        while ix <= len(self._floors):
            floor = self._floors[ix - 1]

            jx = 1
            while jx <= len(floor.rooms()):
                room = floor.rooms()[jx - 1]
                if room.is_occupied():
                    acc = acc + 1

                jx = jx + 1

            ix = ix + 1

        return acc

    def total_rooms_reserved(self):
        acc = 0

        ix = 1
        while ix <= len(self._floors):
            floor = self._floors[ix - 1]

            jx = 1
            while jx <= len(floor.rooms()):
                room = floor.rooms()[jx - 1]
                if room.is_reserved():
                    acc = acc + 1

                jx = jx + 1

            ix = ix + 1

        return acc

    # testing

    def is_empty(self):
        acc = True

        ix = 1
        while ix <= len(self._floors):
            floor = self._floors[ix - 1]

            jx = 1
            while jx <= len(floor.rooms()):
                room = floor.rooms()[jx - 1]
                acc = acc and room.is_available()

                jx = jx + 1

            ix = ix + 1

        return acc

    # guests

    def receive_at_floor_at_room(self, a_guest_type, a_floor_number, a_room_number):
        if a_floor_number < 1 or a_floor_number > len(self._floors):
            Hotel.signal_floor_number_does_not_exist()
        floor = self._floors[a_floor_number - 1]
        if a_room_number < 1 or a_room_number > len(floor.rooms()):
            Hotel.signal_room_number_does_not_exist()
        room = floor.rooms()[a_room_number - 1]

        room.receive(a_guest_type)

        self._available_rooms_count = self._available_rooms_count - 1

    def receive_with_reservation_at_floor_at_room(self, a_guest_type, a_floor_number, a_room_number):
        if a_floor_number < 1 or a_floor_number > len(self._floors):
            Hotel.signal_floor_number_does_not_exist()
        floor = self._floors[a_floor_number - 1]
        if a_room_number < 1 or a_room_number > len(floor.rooms()):
            Hotel.signal_room_number_does_not_exist()
        room = floor.rooms()[a_room_number - 1]

        room.receive_with_reservation(a_guest_type)

    def reserve_room_at_floor(self, a_room_number, a_floor_number):
        if a_floor_number < 1 or a_floor_number > len(self._floors):
            Hotel.signal_floor_number_does_not_exist()
        floor = self._floors[a_floor_number - 1]
        if a_room_number < 1 or a_room_number > len(floor.rooms()):
            Hotel.signal_room_number_does_not_exist()
        room = floor.rooms()[a_room_number - 1]

        room.reserve()

        self._available_rooms_count = self._available_rooms_count - 1

    # assertions (class side)

    @staticmethod
    def assert_have_floors(a_floors_collection):
        if len(a_floors_collection) == 0:
            Hotel.signal_no_floors()

    # error description (class side)

    @staticmethod
    def floor_number_does_not_exist_error_description():
        return "Floor number does not exist"

    @staticmethod
    def no_floors_error_description():
        return "Cannot have a Hotel without floors"

    @staticmethod
    def room_number_does_not_exist_error_description():
        return "Room number does not exist"

    # exceptions (class side)

    @staticmethod
    def signal_floor_number_does_not_exist():
        raise RuntimeError(Hotel.floor_number_does_not_exist_error_description())

    @staticmethod
    def signal_no_floors():
        raise RuntimeError(Hotel.no_floors_error_description())

    @staticmethod
    def signal_room_number_does_not_exist():
        raise RuntimeError(Hotel.room_number_does_not_exist_error_description())
